disco_aws_automation/disco_datapipeline.py

```diff
@@ -0,0 +1,281 @@
+"""
+Back end behaviors for creating data pipelines.
+"""
+
+import json
+from datetime import datetime
+from logging import getLogger
+
+import boto3
+
+from .disco_config import open_normalized
+from .resource_helper import throttled_call
+
+
+_LOG = getLogger(__name__)
+
+
+class DataPipelineConsts(object):
+    """Constants for data pipeline management."""
+    TEMPLATE_DIR = "datapipeline_templates"
+
+
+class AsiaqDataPipeline(object):
+    """Local encapsulation of the information associated with a single data pipeline."""
+    def __init__(self, name, description, tags=None, metadata=None,
+                 pipeline_id=None, contents=None, parameter_definitions=None, param_values=None):
+        self._objects = contents
+        self._params = parameter_definitions
+        self._param_values = _optional_dict_to_list(param_values)
+        self._name = name
+        self._description = description
+        self._tags = _optional_dict_to_list(tags, key_string='key', value_string='value')
+        self._metadata = metadata
+        self._id = pipeline_id
+
+    def is_persisted(self):
+        "Return true if this pipeline has an AWS ID; false otherwise."
+        return self._id is not None
+
+    def has_content(self):
+        "Return true if this pipeline has actual pipeline objects, false if it is only metadata."
+        return self._objects is not None
+
+    def get_tag_dict(self):
+        "Retrieve tags as a dictionary."
+        return _optional_list_to_dict(self._tags, key_string='key', value_string='value')
+
+    def get_param_value_dict(self):
+        "Retrieve parameter values as a dictionary."
+        return _optional_list_to_dict(self._param_values)
+
+    def update_content(self, contents, parameter_definitions, param_values=None):
+        "Set the pipeline content (pipeline nodes, parameters and values) for this pipeline."
+        self._objects = contents
+        self._params = parameter_definitions
+        self._param_values = _optional_dict_to_list(param_values)
+
+    @classmethod
+    def from_template(cls, template_name, name, description, tags=None, param_values=None, log_location=None):
+        """Create a new AsiaqDataPipeline object, populated from template in the configuration directory."""
+        with open_normalized(DataPipelineConsts.TEMPLATE_DIR, template_name + ".json") as f:
+            template_data = json.load(f)
+        boto_objects, boto_parameters = template_to_boto(template_data)
+        if log_location:
+            add_log_location_param(boto_objects, log_location)
+        return cls(contents=boto_objects, parameter_definitions=boto_parameters,
+                   name=name, description=description, tags=tags, param_values=param_values)
+
+
+class AsiaqDataPipelineManager(object):
```

Comment:
Each pipeline object describes the association of a dynamo table and it's s3 file like the sample json file later, right?  Mhh just the name "pipeline" makes me think of the hostclass pipeline.